chia/consensus/mmr.py

```diff
@@ -0,0 +1,279 @@
+        idx -= mountain_size
+
+    return peaks
+
+
+def leaf_index_to_pos(leaf_index: int) -> int:
+    """
+    Convert a leaf index (0-based) to its position in the flat MMR.
+    Formula: 2*L - popcount(L)
+    """
+    return 2 * leaf_index - bin(leaf_index).count("1")
+
+
+# ------------------------------------------------------------------------------
+# Class Implementation
+# ------------------------------------------------------------------------------
+
+
+class MerkleMountainRange(Streamable):
+    """
+    Flat MMR implementation.
+    """
+
+    nodes: list[bytes32]
+    size: uint32
+
+    def __init__(self, nodes: list[bytes32] | None = None, size: uint32 = uint32(0)) -> None:
+        self.nodes = [] if nodes is None else nodes
+        self.size = size
+
+    def append(self, leaf: bytes32) -> None:
+        nodes = self.nodes
+        curr_index = len(nodes)
+        nodes.append(leaf)
+
+        curr_height = 0
+
+        # Merge upwards
+        while True:
+            # Size of subtree at current height: 2^(h+1) - 1
+            subtree_size = (1 << (curr_height + 1)) - 1
+
+            # Potential left sibling is 'subtree_size' back
+            left_sibling_index = curr_index - subtree_size
+
+            if left_sibling_index < 0:
+                break
+
+            # If left node has same height, merge
+            if get_height(left_sibling_index) == curr_height:
+                left_hash = nodes[left_sibling_index]
+                right_hash = nodes[curr_index]
+
+                parent_hash = std_hash(left_hash + right_hash)
+
+                # Append parent
+                nodes.append(parent_hash)
+
+                # Move focus to the new parent
+                curr_index = len(nodes) - 1
+                curr_height += 1
+            else:
+                # Different heights means we started a new mountain - stop merging
+                break
+
+        self.size = uint32(self.size + 1)
+        log.debug(f"Appended new leaf, MMR size is now {self.size}, total nodes: {len(self.nodes)}")
+
+    def get_root(self) -> bytes32 | None:
+        """Get the MMR root by bagging the peaks."""
+        peak_indices = get_peak_positions(len(self.nodes))
+        if not peak_indices:
+            return None
+
+        # Bagging Order: Rightmost (Smallest) -> Leftmost (Tallest)
+        current_hash = self.nodes[peak_indices[0]]
+
+        for i in range(1, len(peak_indices)):
+            left_peak = self.nodes[peak_indices[i]]
+            current_hash = std_hash(left_peak + current_hash)
+
+        return current_hash
+
+    def get_tree_height(self) -> int:
+        if self.size == 0:
+            return 0
+        peak_indices = get_peak_positions(len(self.nodes))
+        if not peak_indices:
+            return 0
+        return get_height(peak_indices[-1])
+
+    def copy(self) -> MerkleMountainRange:
+        return MerkleMountainRange(list(self.nodes), uint32(self.size))
```

Comment:
in a world where you don't use checkpoints, but just roll back the MMR in-place, I don't think you need this. For the augmented chain you would need something similar though, that would return a slice/subset of the MMR, probably by copying the slice/subset, to allow the augmented chain to add different blocks to it